File: solutions/R/mixed_models_ex2.R
# =============================================================================
# Chapter 6b, Exercise 2: Fit and interpret a random-intercept model
# Recreate the BP data, add a binary treatment, fit a random-intercept LMM,
# compute the ICC, and compare the treatment SE against ordinary lm().
# =============================================================================

library(tidyverse)   # data wrangling and simulation
library(lme4)        # lmer() for linear mixed models
# lmerTest adds p-values to lmer(); use it if available, otherwise fall back
# to estimates + profile confidence intervals from base lme4.
has_lmerTest <- requireNamespace("lmerTest", quietly = TRUE)
if (has_lmerTest) suppressMessages(library(lmerTest))

set.seed(42)

# --- Recreate the chapter's longitudinal, clustered BP dataset --------------
n_clinic  <- 8     # 8 clinics
n_patient <- 25    # 25 patients per clinic
n_visit   <- 5     # 5 visits per patient

clinic_effect  <- rnorm(n_clinic, 0, 6)              # clinic baselines (SD 6)
patient_effect <- rnorm(n_clinic * n_patient, 0, 8)  # patient baselines (SD 8)

# Add a simulated binary treatment, assigned at the PATIENT level (a patient is
# either treated or not, and keeps that assignment across all their visits).
# True treatment effect built into the simulation: -5 mmHg.
treatment_by_patient <- rbinom(n_clinic * n_patient, 1, 0.5)

bp <- expand_grid(
  clinic  = 1:n_clinic,
  patient = 1:n_patient,
  visit   = 1:n_visit
) |>
  mutate(
    patient_id = interaction(clinic, patient),   # unique patient label
    pid_index  = as.integer(patient_id),
    treatment  = treatment_by_patient[pid_index],
    sbp = 135 +
          clinic_effect[clinic] +
          patient_effect[pid_index] +
          -1.5 * (visit - 1) +      # downward drift per visit
          -5   * treatment +        # true treatment effect: -5 mmHg
          rnorm(n(), 0, 5)          # residual noise
  )

cat("Dataset:", nrow(bp), "rows |",
    n_clinic, "clinics x", n_patient, "patients x", n_visit, "visits\n\n")

# --- Fit the random-intercept model -----------------------------------------
# Fixed effects: visit + treatment. Random intercepts for clinic and for
# patient nested within clinic.
m_ri <- lmer(sbp ~ visit + treatment + (1 | clinic/patient_id), data = bp)

cat("=== Random-intercept model summary ===\n")
print(summary(m_ri))

# --- (a) Treatment fixed effect and its interpretation ----------------------
coefs   <- summary(m_ri)$coefficients
beta_trt <- coefs["treatment", "Estimate"]
se_trt   <- coefs["treatment", "Std. Error"]

cat("\n=== (a) Treatment fixed effect ===\n")
cat(sprintf("Estimate: %.3f mmHg   SE: %.3f\n", beta_trt, se_trt))
if (has_lmerTest && "Pr(>|t|)" %in% colnames(coefs)) {
  cat(sprintf("p-value (lmerTest): %.4g\n", coefs["treatment", "Pr(>|t|)"]))
} else {
  ci <- confint(m_ri, parm = "treatment", method = "Wald")
  cat(sprintf("95%% CI (Wald): %.3f to %.3f\n", ci[1], ci[2]))
}
cat("Interpretation: after accounting for visit and the clustering of visits\n")
cat("  within patients and patients within clinics, being on treatment is\n")
cat(sprintf(
  "  associated with a %.2f mmHg change in systolic BP (a reduction),\n",
  beta_trt))
cat("  holding visit number constant.\n")

# --- (b) Intraclass correlation ---------------------------------------------
vc <- as.data.frame(VarCorr(m_ri))
var_clinic  <- vc$vcov[vc$grp == "clinic"]
var_patient <- vc$vcov[vc$grp == "patient_id:clinic"]
var_resid   <- vc$vcov[vc$grp == "Residual"]
var_total   <- var_clinic + var_patient + var_resid

icc_clinic  <- var_clinic / var_total
icc_patient <- var_patient / var_total
# Proportion of variance at or above the patient level (clinic + patient):
icc_cluster <- (var_clinic + var_patient) / var_total

cat("\n=== (b) Variance components and ICC ===\n")
cat(sprintf("Between-clinic variance : %.2f\n", var_clinic))
cat(sprintf("Between-patient variance: %.2f\n", var_patient))
cat(sprintf("Residual variance       : %.2f\n", var_resid))
cat(sprintf("Clinic-level ICC : %.3f\n", icc_clinic))
cat(sprintf("Patient-level ICC: %.3f\n", icc_patient))
cat(sprintf(
  "Plain English: about %.0f%% of the total variation in blood pressure lies\n",
  100 * icc_cluster))
cat(sprintf(
  "  between clinics and patients rather than within a patient's own visits\n"))
cat(sprintf(
  "  (clinic alone accounts for roughly %.0f%%).\n", 100 * icc_clinic))

# --- (c) Compare with ordinary lm() that ignores clustering -----------------
m_lm <- lm(sbp ~ visit + treatment, data = bp)
lm_coefs <- summary(m_lm)$coefficients
se_trt_lm <- lm_coefs["treatment", "Std. Error"]

cat("\n=== (c) Ordinary lm() ignoring clustering ===\n")
cat(sprintf("Treatment estimate (lm)  : %.3f\n", lm_coefs["treatment", "Estimate"]))
cat(sprintf("Treatment SE (lm)        : %.3f\n", se_trt_lm))
cat(sprintf("Treatment SE (mixed)     : %.3f\n", se_trt))
cat(sprintf("Ratio mixed/lm           : %.2f\n", se_trt / se_trt_lm))
cat("\nWhich is larger, and why it matters:\n")
cat("  The MIXED-model SE is larger. Treatment is a BETWEEN-PATIENT variable\n")
cat("  (constant across a patient's 5 visits), so the real amount of\n")
cat("  independent information about treatment is roughly the number of\n")
cat("  patients, not the 1000 rows. Ordinary lm() pretends all 1000 rows are\n")
cat("  independent, so it reports an SE that is too small -- an over-optimistic\n")
cat("  confidence interval and p-value. Ignoring the clustering would make the\n")
cat("  treatment effect look more certain than the data can support.\n")

File: solutions/python/mixed_models_ex2.py
# =============================================================================
# Chapter 6b, Exercise 2: Fit and interpret a random-intercept model
# Recreate the BP data, add a binary treatment, fit a random-intercept MixedLM,
# compute the ICC, and compare the treatment SE against ordinary OLS.
# =============================================================================

import numpy as np
import pandas as pd
import statsmodels.formula.api as smf

np.random.seed(42)

# --- Recreate the chapter's longitudinal, clustered BP dataset (long format) -
n_clinic, n_patient, n_visit = 8, 25, 5
clinic_effect = np.random.normal(0, 6, n_clinic)   # clinic baselines (SD 6)

rows = []
pid = 0
for c in range(n_clinic):
    for p in range(n_patient):
        patient_effect = np.random.normal(0, 8)        # patient baseline (SD 8)
        # Treatment assigned at the PATIENT level (constant across the patient's
        # visits). True built-in effect: -5 mmHg.
        treatment = np.random.binomial(1, 0.5)
        for v in range(n_visit):
            sbp = (135 + clinic_effect[c] + patient_effect
                   - 1.5 * v            # downward drift per visit
                   - 5.0 * treatment    # true treatment effect: -5 mmHg
                   + np.random.normal(0, 5))
            rows.append({"clinic": c, "patient_id": pid, "visit": v,
                         "treatment": treatment, "sbp": sbp})
        pid += 1

bp = pd.DataFrame(rows)
print(f"Dataset: {len(bp)} rows | {n_clinic} clinics x {n_patient} patients "
      f"x {n_visit} visits\n")

# --- Fit the random-intercept model -----------------------------------------
# statsmodels MixedLM groups by ONE level; we use a random intercept per
# patient (the level at which treatment varies), which absorbs the repeated-
# visit correlation. Fixed effects: visit + treatment.
m_ri = smf.mixedlm("sbp ~ visit + treatment", data=bp, groups=bp["patient_id"])
res_ri = m_ri.fit()

print("=== Random-intercept model summary ===")
print(res_ri.summary())

# --- (a) Treatment fixed effect and its interpretation ----------------------
beta_trt = res_ri.params["treatment"]
se_trt = res_ri.bse["treatment"]
p_trt = res_ri.pvalues["treatment"]

print("\n=== (a) Treatment fixed effect ===")
print(f"Estimate: {beta_trt:.3f} mmHg   SE: {se_trt:.3f}   p-value: {p_trt:.4g}")
print("Interpretation: after accounting for visit and the fact that repeated")
print("  visits are clustered within patients, being on treatment is")
print(f"  associated with a {beta_trt:.2f} mmHg change in systolic BP (a")
print("  reduction), holding visit number constant.")

# --- (b) Intraclass correlation ---------------------------------------------
# ICC at the patient level = between-patient variance / total variance.
var_group = float(res_ri.cov_re.iloc[0, 0])   # between-patient variance
var_resid = float(res_ri.scale)               # residual (within-patient) var
icc = var_group / (var_group + var_resid)

print("\n=== (b) Variance components and ICC ===")
print(f"Between-patient variance: {var_group:.2f}")
print(f"Residual variance       : {var_resid:.2f}")
print(f"Patient-level ICC       : {icc:.3f}")
print(f"Plain English: about {100*icc:.0f}% of the total variation in blood")
print("  pressure is due to persistent differences between patients rather than")
print("  visit-to-visit fluctuation within a patient.")

# --- (c) Compare with ordinary OLS that ignores clustering ------------------
m_ols = smf.ols("sbp ~ visit + treatment", data=bp).fit()
se_trt_ols = m_ols.bse["treatment"]

print("\n=== (c) Ordinary OLS ignoring clustering ===")
print(f"Treatment estimate (OLS): {m_ols.params['treatment']:.3f}")
print(f"Treatment SE (OLS)      : {se_trt_ols:.3f}")
print(f"Treatment SE (mixed)    : {se_trt:.3f}")
print(f"Ratio mixed/OLS         : {se_trt / se_trt_ols:.2f}")
print("\nWhich is larger, and why it matters:")
print("  The MIXED-model SE is larger. Treatment is a BETWEEN-PATIENT variable")
print("  (constant across a patient's 5 visits), so the real amount of")
print("  independent information about treatment is roughly the number of")
print("  patients, not the 1000 rows. Ordinary OLS pretends all 1000 rows are")
print("  independent, so it reports an SE that is too small -- an over-")
print("  optimistic confidence interval and p-value. Ignoring the clustering")
print("  would make the treatment effect look more certain than the data")
print("  can support.")
